### PR TITLE
Fix ABI checking...

### PR DESCRIPTION
**Pull request by mtl1979**
* Ubuntu 22.04 use different format for ABI files so old ones need to be removed
* Use more recent zlib-ng commit to avoid issues with internal adler32 and crc32 functions

### CODE REVIEW

**Comment by nmoinvaz** [comment]
> When new ABI files are uploaded, Ubuntu 22.04 must be used to create them, and CI runs needs to be locked to Ubuntu 22.04.

ABI files can be found in any Package Check workflow that fails; they are uploaded as artifacts. For example, here is from a recent run (older runs may not have retained the artifacts):
https://github.com/zlib-ng/zlib-ng/actions/runs/3790538662

**Comment by mtl1979** [comment]
@nmoinvaz I was waiting for GitHub to upgrade all "latest" workflows to 22.04 first... Otherwise the non-upgraded ones will fail instead due to file format inconsistency.

**Comment by mtl1979** [comment]
Rebased only to include recent fixes to CI runs...

**Comment by nmoinvaz** [comment]
@mtl1979 perhaps all the machines have been updated now?

**Comment by mtl1979** [comment]
> @mtl1979 perhaps all the machines have been updated now?

There is still pull requests that can change our API. Kind of chicken and egg problem.

**Comment by mtl1979** [comment]
Rebased again and fixed issue with visibility of `crc32_pclmulqdq`.

**Comment by mtl1979** [comment]
@nmoinvaz Now there is should be quite nice amount of ABI files...

**Comment by nmoinvaz** [comment]
Wow so much green checkmarks! Only GCC is broken but that is a separate issue where perhaps `<xsaveintrin.h>` needs to be included.

**Comment by mtl1979** [comment]
@nmoinvaz I made a script that tries to build zlib and zlib-ng for all possible targets... Running the script takes ages, but it saves all the ABI files automatically ;)

**Comment by mtl1979** [comment]
@nmoinvaz MinGW build was completely broken... I added separate commit so I can test if we get all greens...

**Comment by nmoinvaz** [comment]
See my comment here: https://github.com/zlib-ng/zlib-ng/pull/1407#discussion_r1082970732. Seems it was recently introduced.

File: cmake/detect-intrinsics.cmake [hunk 1]
```diff
@@ -536,3 +536,21 @@ macro(check_vgfma_intrinsics)
         HAVE_VGFMA_INTRIN FAIL_REGEX "not supported")
     set(CMAKE_REQUIRED_FLAGS)
 endmacro()
+
+macro(check_xsave_intrinsics)
+    if(NOT NATIVE_FLAG)
```
Comment (nmoinvaz) [inline cmake/detect-intrinsics.cmake]: Should be `if (NOT NATIVEFLAG AND NOT MSVC)`? Notice extra underscore you added in `NATIVEFLAG`.

File: configure [hunk 2]
```diff
@@ -1224,6 +1241,27 @@ EOF
...
+#else
+#  include <x86gprintrin.h>
+#endif
+int main(void) {
+    return _xgetbv(0);
+}
+EOF
+    if try ${CC} ${CFLAGS} ${xsaveflag} $test.c; then
+        echo "Checking for XSAVE intrinsics ... Yes." | tee -a configure.log
+        HAVE_XSAVE_INTRIN=1
```
Comment (nmoinvaz) [inline configure]: This `HAVE_XSAVE_INTRIN` does not appear to be used anywhere.

File: cmake/detect-intrinsics.cmake (hunk 1 above)
Comment (mtl1979) [reply cmake/detect-intrinsics.cmake]: Possibly... On MSVC it would just fail and not set the flag...

File: configure (hunk 2 above)
Comment (mtl1979) [reply configure]: Fixed.

File: cmake/detect-intrinsics.cmake (hunk 1 above)
Comment (nmoinvaz) [reply cmake/detect-intrinsics.cmake]: We can just return() on MSVC then.

**Comment by mtl1979** [comment]
> See my comment here: [#1407 (comment)](https://github.com/zlib-ng/zlib-ng/pull/1407#discussion_r1082970732). Seems it was recently introduced.

There was other errors with MinGW build too... Basically related to symbol mangling, gzip functions and out-of-tree builds. I got a lot of linker errors while trying to get XSAVE test working with configure.

File: cmake/detect-intrinsics.cmake (hunk 1 above)
Comment (mtl1979) [reply cmake/detect-intrinsics.cmake]: It doesn't hurt to let MSVC test.... Would not be the first time Microsoft messes up with the intrinsics headers and breaks things... One improvement would be to add preprocessor flag so that on non-Windows platforms, it would also use the intrinsic function and not inline assembly... But some gcc versions are known to have buggy intrinsic function...

**Comment by nmoinvaz** [comment]
I think perhaps the xsave stuff should be its own PR.

File: CMakeLists.txt [hunk 3]
```diff
@@ -740,9 +740,14 @@ if(WITH_OPTIM)
             endif()
         endif()
     elseif(BASEARCH_X86_FOUND)
+        check_xsave_intrinsics()
```
Comment (nmoinvaz) [inline CMakeLists.txt]: I think this entire XSave block should go at the end of the if statement. I've tried to keep all the checks alphabetical.

Comment (nmoinvaz) [reply CMakeLists.txt]: But idk, it might be ok as is. Maybe put fallback_bultin's block above this one?

File: configure [hunk 4]
```diff
@@ -452,12 +460,16 @@ if test "$gcc" -eq 1 && ($cc $CFLAGS -c $test.c) >> configure.log 2>&1; then
         LDSHARED=${LDSHARED-"$cc"}
         LDSHAREDFLAGS="-shared -Wl,--out-implib,${IMPORTLIB}"
         LDSHAREDLIBC=""
-        DEFFILE='win32/${LIBNAME2}.def'
+        if test $gzfileops -eq 0; then
```
Comment (nmoinvaz) [inline configure]: I'm not sure what some of these visibility, RCFLAGS, gzfileops changes are. Perhaps they should be separate commits.

File: configure [hunk 5]
```diff
@@ -1500,6 +1542,12 @@ case "${ARCH}" in
             CFLAGS="${CFLAGS} -DX86_FEATURES"
             SFLAGS="${SFLAGS} -DX86_FEATURES"
 
+            check_xsave_intrinsics
```
Comment (nmoinvaz) [inline configure]: This should definitely go in alphabetical order, at the bottom of this if statement.

File: CMakeLists.txt (hunk 3 above)
Comment (nmoinvaz) [reply CMakeLists.txt]: Yeah I think it should be alphabetical, in the same order as configure script.

Comment (mtl1979) [reply CMakeLists.txt]: This has to be before line 749, but otherwise the placement isn't critical...

File: configure (hunk 4 above)
Comment (mtl1979) [reply configure]: MinGW doesn't support visibility attribute, but gives only warning, not error, so it makes sense to not check it for configure as we don't parse the compiler output like cmake does.

`RCFLAGS` adds binary/output directory to the include path for `windres` as it doesn't do it automatically. Otherwise it won't find `zlib.h` or `zlib-ng.h`.

File: configure (hunk 5 above)
Comment (mtl1979) [reply configure]: I added it here, because it's only relevant to `x86_features.h`. But I can move it...

**Comment by mtl1979** [comment]
> I think perhaps the xsave stuff should be its own PR.

Before, I made separate pull requests for related patches, but this whole PR grew up bigger than I planned due to adding the ABI files and revealing all the build failures behind the noise.

`xsave` tests and all the MinGW build fixes have to be on same PR, otherwise it will be hard to verify that there is no more errors as the build will not finish successfully without including all the changes.

All the MinGW fixes are now separated as #1412. It can only be merged after this PR, so I marked it as a draft.

**Comment by mtl1979** [comment]
@Dead2 Last chance to comment... I want this merged as fast as possible so we can merge all the PRs that depend on ABI checks working correctly.

**Review by nmoinvaz — APPROVED** [review]
LGTM.

**Review by Dead2 — APPROVED** [review]
LGTM

**Comment by mtl1979** [comment]
Rebasing due to 1ab443812aceab6c80ccd573ecdd06f73a84eef4 and generating new ABI files.
